In [ ]:
# !pip install openai
# !pip install python-dotenv

In [9]:
from dotenv import load_dotenv
import os
from openai import OpenAI

In [59]:
load_dotenv()

True

In [11]:
print("Loaded environment variables.")
print(os.getenv("OPENAI_API_KEY"))

Loaded environment variables.
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


In [49]:
#client = OpenAI()
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

In [13]:
response = client.responses.create(
    model="gpt-5.1",
    input="Write a short bedtime story about a unicorn."
)
print(response.output_text)


In a quiet valley where the stars shone brighter than anywhere else, there lived a small unicorn named Liora. Her coat was soft and silver, and her horn glowed with the palest blue light, like moonlight on snow.

Every unicorn in the valley had a special gift. Some could make flowers bloom with a stomp of their hooves, others could sing songs that made rainbows appear. Liora, however, did not yet know what her gift was.

Each night, she watched the older unicorns practice their magic. She tried to help: she stomped her hooves over the grass, but no flowers came. She sang softly to the sky, but no rainbows formed. She dipped her horn into the river, hoping it would turn to sparkling sugar, but the water stayed water.

One evening, feeling a little sad, Liora wandered away from the valley and climbed a small hill. Above her, the sky was dark and deep, sprinkled with thousands of tiny stars. The moon hung low and round, like a silver lantern.

“I wish I knew what I’m meant to do,” she whi

In [ ]:
response = client.responses.create(
    model="gpt-5.1",
    input="Write a short bedtime story about a unicorn.",
    temperature=0.7, # Creativity Control : 0.0 (more focused) to 1.0 (more creative)
    top_p=1.0, # Nucleus Sampling: 0.0 to 1.0: 1.0 means consider all tokens : 0.5 means consider only top 50% probable tokens
    max_output_tokens=200,
    frequency_penalty=0.3, # Discourage Repetition: 0.0 to 2.0
    presence_penalty=0.0 # Encourage New Topics: 0.0 to 2.0
)

TypeError: Responses.create() got an unexpected keyword argument 'frequency_penalty'

In [53]:
## Use Chat Completion not Chat Response
from openai import OpenAI
client = OpenAI()

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[{"role": "user", "content": "Write a bedtime story about a unicorn"}],
    temperature=0.7,
    top_p=1.0,
    frequency_penalty=0.5,
    presence_penalty=0.3
)




In [52]:
print(response.choices[0].message.content)

**The Moonlit Journey of Luna the Unicorn**

Once upon a time in a mystical land called Celestia, where the skies shimmered with rainbow colors and the trees whispered secrets to one another, there lived a gentle unicorn named Luna. She had a coat as white as freshly fallen snow and a mane that sparkled like stardust. But what made Luna truly special was her magnificent horn, which glowed softly in the moonlight, casting a silver light over everything around her.

Luna loved to wander through the enchanted forest that surrounded her home. Each evening, as the sun dipped below the horizon and painted the sky with hues of orange and pink, she would set off on adventures, exploring hidden meadows filled with flowers that danced in the breeze.

One evening, while wandering deeper into the forest than she had ever been before, Luna stumbled upon an ancient oak tree with twisting roots and shimmering leaves. As she approached, she heard a soft sobbing sound coming from beneath its branches. 

## Test Transcripts

In [1]:
from openai import OpenAI
client = OpenAI()

audio_file = open("speech_test_regenerated.mp3", "rb")

transcript = client.audio.transcriptions.create(
    model="gpt-4o-transcribe",
    file=audio_file
)

print(transcript)


Transcription(text='Hi there, this is a regenerated audio sample for testing the OpenAI transcription API. This audio is short, clear, and suitable for your transcription experiments.', logprobs=None, usage=UsageTokens(input_tokens=88, output_tokens=33, total_tokens=121, type='tokens', input_token_details=UsageTokensInputTokenDetails(audio_tokens=88, text_tokens=0)))


# GROQ

### Set the key in .env
### load env

In [4]:
#!pip install groq

In [7]:
from dotenv import load_dotenv
import os
load_dotenv()

True

In [60]:
from groq import Groq

client = Groq()
completion = client.chat.completions.create(
    model="llama-3.1-8b-instant",
    messages=[
      {
        "role": "user",
        "content": ""
      }
    ],
    temperature=1,
    max_completion_tokens=1024,
    top_p=1,
    stream=True,
    stop=None
)

for chunk in completion:
    print(chunk.choices[0].delta.content or "", end="")


Is there something I can help you with?

In [9]:
client = Groq()
completion = client.chat.completions.create(
    model="llama-3.1-8b-instant",
    messages=[
      {
        "role": "user",
        "content": "Share the latest news about AI research from the past month."
      }
    ],
    temperature=1,
    max_completion_tokens=1024,
    top_p=1,
    stream=True,
    stop=None
)

for chunk in completion:
    print(chunk.choices[0].delta.content or "", end="")


Since my knowledge cutoff is December 2023, I'll provide information on AI research news from the period leading up to that date. However, please note that some of the information might have changed or been updated since then.

Here are some of the latest developments in AI research from the past few months (up to December 2023):

1. **Breakthroughs in Large Language Models:** Researchers have made significant progress in developing more efficient and accurate large language models (LLMs). For example, Meta AI's LLaMA model has shown excellent performance in various NLP tasks, and the researchers have also made improvements to the model architecture.

2. **Advances in Explainable AI (XAI):** Researchers at Google have made strides in developing more interpretable and explainable AI models. They have created a new tool, Explainable Neural Networks (ENN), which can provide human-interpretable explanations for AI decisions.

3. **AI for Cybersecurity:** Researchers at the University of Ca

In [11]:
!pip install ollama

# deepseek-r1:1.5b Using OLlama

In [15]:
from ollama import chat
from ollama import ChatResponse

response: ChatResponse = chat(model='deepseek-r1:1.5b', messages=[
  {
    'role': 'user',
    'content': 'Why is the sky blue?',
  },
])
#print(response['message']['content'])
# or access fields directly from the response object
print(response.message.content)

The color of the sky, known as coloration or chromacy, is primarily determined by the amount of sunlight reflected on the Earth's surface. Different parts of the sky have varying levels of this reflection because light reflecting off water particles in the atmosphere, such as clouds and mist, can pass through more air molecules than light from objects like buildings or cars that reflects off dust or other particles. This difference in reflection leads to slight variations in the hues we perceive when looking up at the sky. Additionally, atmospheric conditions, including temperature, humidity, and the presence of trace gases like carbon dioxide, influence how the sky appears.


# Using Langchain

In [18]:
#!pip install langchain-openai

In [19]:
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate

llm = ChatOpenAI(model="gpt-4o-mini")

prompt = ChatPromptTemplate.from_messages([
    ("system", "You are a helpful assistant."),
    ("human", "{question}")
])

chain = prompt | llm
print(chain.invoke({"question": "What is Agentic AI?"}))


content='Agentic AI refers to artificial intelligence systems that possess the ability to take autonomous actions, make decisions, and interact with the environment to achieve specific goals or tasks. Unlike traditional AI, which often operates under human direction and focuses on tasks like data analysis or pattern recognition, agentic AI can operate independently and adaptively.\n\nKey characteristics of agentic AI may include:\n\n1. **Autonomy**: The AI can perform tasks and make decisions without human intervention.\n2. **Goal-oriented behavior**: It is designed to achieve specific objectives or goals, which could involve planning and decision-making.\n3. **Adaptability**: Agentic AI can learn from its experiences and change its behavior based on new information or feedback from its environment.\n4. **Interactivity**: These systems can often interact with complex environments or with humans and other agents to achieve their objectives.\n\nExamples of agentic AI can range from advan

In [22]:
#pip install langchain-groq

In [61]:
from langchain_groq import ChatGroq
llm = ChatGroq(model="llama-3.1-8b-instant")

response = llm.invoke("Explain transformers in simple words.")
print(response)


content="Transformers are a type of machine that can change the voltage of an electrical current. They are commonly used in power grids to make it easier to transmit electricity over long distances.\n\nHere's a simple analogy to help explain how transformers work:\n\nImagine you have a big pipe that carries water from a river to a city. The pipe is like a power line that carries electricity from a power plant to a city. But the problem is that the water (or electricity) is flowing too fast and would be too strong for the city's homes and businesses to handle.\n\nA transformer is like a special valve that can slow down the water flow (or reduce the voltage) so that it's safe and usable for the city's homes and businesses. This way, the city can use the electricity without it being too strong or too weak.\n\nThere are two types of transformers:\n\n1. **Step-up transformer**: This type of transformer increases the voltage of the electrical current, like a pump that makes the water flow fa

## System Prompt

In [25]:
from langchain_core.prompts import PromptTemplate

template = PromptTemplate(
    input_variables=["topic"],
    template="Explain {topic} in 3 bullet points."
)

print(template.format(topic="embeddings"))


Explain embeddings in 3 bullet points.


In [26]:
from langchain_core.output_parsers import JsonOutputParser
from langchain_openai import ChatOpenAI

parser = JsonOutputParser()

prompt = """
Extract company name and revenue from the text below.
Return output ONLY in JSON format with keys company, revenue.

Text:
Apple generated $383B revenue in 2023.
"""

model = ChatOpenAI()
print(parser.parse(model.invoke(prompt).content))


{'company': 'Apple', 'revenue': '$383B'}


# Embeddings

In [27]:
from openai import OpenAI

client = OpenAI()

# Example text list
texts = [
    "Retrieval-Augmented Generation (RAG) improves accuracy.",
    "Embeddings convert text into high-dimensional vectors."
]

response = client.embeddings.create(
    model="text-embedding-3-small",
    input=texts
)

# Print embeddings
for i, emb in enumerate(response.data):
    print(f"Text {i+1} embedding length:", len(emb.embedding))
    print(emb.embedding[:10], "...")  # show first 10 numbers


Text 1 embedding length: 1536
[0.015356769785284996, 0.03452569618821144, 0.03314683213829994, 0.01639767736196518, -0.021115558221936226, -0.0155595438554883, 0.013437173329293728, 0.012734223157167435, -0.018749859184026718, 0.019087815657258034] ...
Text 2 embedding length: 1536
[-0.027849717065691948, 0.003486940171569586, 0.04356098920106888, 0.0014729317044839263, 0.014623391442000866, -0.01214989647269249, 0.015505146235227585, -0.00042477401439100504, 0.002828486729413271, 0.023143205791711807] ...


In [54]:
response = client.embeddings.create(
    model="text-embedding-3-small",
    input=texts,
    dimensions=512
)

# Print embeddings
for i, emb in enumerate(response.data):
    print(f"Text {i+1} embedding length:", len(emb.embedding))
    print(emb.embedding[:10], "...")  # show first 10 numbers


Text 1 embedding length: 512
[-0.006820265203714371, 0.01068376749753952, -0.022077158093452454, 0.030908022075891495, -0.06736687570810318, -0.03038763254880905, 0.06591609120368958, 0.033778052777051926, -0.00962721835821867, -0.008815093897283077] ...
Text 2 embedding length: 512
[-0.03966500610113144, -0.028670484200119972, 0.03462023660540581, -0.05522293224930763, 0.05333595722913742, -0.043862566351890564, -0.04640420898795128, -0.005848662927746773, -0.045826561748981476, 0.06434973329305649] ...
Text 3 embedding length: 512
[-0.005087872035801411, 0.039913903921842575, 0.016241736710071564, -0.00626326072961092, -0.004804299212992191, -0.030938206240534782, 0.03853302821516991, -0.001427111099474132, -0.020976169034838676, 0.027683284133672714] ...


#### Size
    1. text-embedding-3-large : Default vector size: 3072 dimensions [Large]
    2. text-embedding-3-small : 1536 dimensions
    

In [30]:
#!pip install langchain-community

In [31]:
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import FAISS




In [32]:
# Step 1. Embedding model
embedding_model = OpenAIEmbeddings(model="text-embedding-3-small")

# Step 2. Sample documents
docs = [
    "India won its Test match by 200 runs.",
    "RAG improves retrieval accuracy for enterprise systems.",
    "Transfer learning speeds up deep learning training."
]

# Step 3. Create vector store
faiss_index = FAISS.from_texts(docs, embedding_model)



Top results:
- Transfer learning speeds up deep learning training.
- RAG improves retrieval accuracy for enterprise systems.


In [33]:
# Step 4. Search
query = "How do I improve AI accuracy?"
results = faiss_index.similarity_search(query, k=2)

print("Top results:")
for r in results:
    print("-", r.page_content)

Top results:
- Transfer learning speeds up deep learning training.
- RAG improves retrieval accuracy for enterprise systems.


In [34]:
db = FAISS.from_texts(docs, embedding_model)

In [35]:
# Save it
db.save_local("faiss_index")

In [36]:
# Load existing FAISS index
db = FAISS.load_local("faiss_index", embedding_model, allow_dangerous_deserialization=True)


In [37]:
docs_new = [
    "AI can automate repetitive tasks.",
    "Machine learning models are improving healthcare outcomes.",
    "The stock market crashed yesterday due to global tension.",
    "Cats are very playful animals.",
    "Python is widely used in data science.",
    "Electric vehicles are becoming more popular.",
    "The weather today is sunny with mild winds.",
    "Football is the most popular sport in the world.",
    "Quantum computing is the future of cryptography.",
    "Deep learning helps in image recognition.",
    "The restaurant serves Italian and Mexican food.",
    "Traveling to Japan is on my wishlist.",
    "Reinforcement learning trains agents via rewards.",
    "Natural Language Processing enables chatbots.",
    "The new iPhone was launched with major upgrades.",
    "Meditation improves mental health.",
    "Blockchain technology ensures secure transactions.",
    "Yoga enhances flexibility and strength.",
    "Cloud computing allows scalable applications.",
    "Climate change impacts global ecosystems."
]

In [38]:
db.add_texts(docs_new)

['fedbf86f-e0f3-4aa6-82a8-9438d174773f',
 '64fba312-22a9-440d-921e-1c029df9593d',
 '892017d6-430c-4215-abfc-e7dd386749f9',
 'bbb04f95-0d53-4a0c-b4b0-dd6813cac7c9',
 'dedf16ef-62e1-46af-b8b4-b0a663ca0d31',
 'adc02ff3-8444-4583-9306-ab0a636b2f56',
 '55cfc178-a97c-41a2-9eff-d259ba1d4613',
 '5e6cb105-5166-4390-8617-86f0ffcfed28',
 '59cb40b5-4b3e-4d25-a79d-b95128d3afed',
 '5a585964-b038-4b3b-8a7d-69775f19282d',
 '511f5dd6-d287-42c1-a5ab-facff615d06f',
 '31b7c22b-d3ee-486f-aa7e-dc7f1710ef68',
 'ecf6ba3d-a2e2-478f-a817-a4dc98ac4e27',
 '23b4fdaf-88b3-4977-bc0e-34d03675b16a',
 '2d26d37c-bcb1-40cc-9523-cd000ecea548',
 'c3158f71-c5aa-4f3f-99e8-f4661c5e8e8d',
 '4c805f06-82ca-4078-ae92-e320e5edbaff',
 'b0c43edd-4a34-4866-9bea-5e0594b92177',
 'abe38cf7-043e-470a-84c2-ed29fa58ddf6',
 '0a0be554-5325-468e-a1c5-0f40917ec87e']

In [39]:
# Save updated index
db.save_local("faiss_index")

In [41]:
query = "How can AI help in healthcare?"
results = db.similarity_search(query, k=3)
print("\nTop search results:")
for r in results:
    print("-", r.page_content)


Top search results:
- Machine learning models are improving healthcare outcomes.
- AI can automate repetitive tasks.
- Deep learning helps in image recognition.


In [42]:
query = "How do vector databases work?"

results = db.similarity_search_with_score(query, k=3)

for doc, score in results:
    print("Text:", doc.page_content)
    print("Score:", score)
    print("---")


Text: RAG improves retrieval accuracy for enterprise systems.
Score: 1.3513801
---
Text: Cloud computing allows scalable applications.
Score: 1.5137949
---
Text: Transfer learning speeds up deep learning training.
Score: 1.5689678
---


In [43]:
print("FAISS index type:", db.index.metric_type)


FAISS index type: 1


#### distance
1 -- Inner Product (higher is better)

0 -- L2 Distance (Smaller is better)

In [45]:
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import FAISS
from langchain_community.vectorstores.utils import DistanceStrategy

emb = OpenAIEmbeddings(model="text-embedding-3-small")

texts = [
    "India won the cricket match.",
    "FAISS is used for similarity search.",
    "Embeddings turn text into vectors."
]

db = FAISS.from_texts(
    texts,
    emb,
    distance_strategy=DistanceStrategy.COSINE
)

print("Metric type:", db.index.metric_type)


Metric type: 1


#### Different Distance
    1. DistanceStrategy.EUCLIDEAN (L2)
        - L2
        - Lower = closer
        - Useful for numeric vectors, clustering
    2. DistanceStrategy.COSINE
        - Equivalent to cosine similarity
        - Higher = closer
        - Best for RAG, semantic search
    3. DistanceStrategy.MAX_INNER_PRODUCT
        - Higher = similar
        - Good for ranking tasks
        - Recommendation Systems (weighted embeddings), Product Ranking on E-commerce Platforms 


In [ ]:
from langchain_community.vectorstores.utils import DistanceStrategy

# Cosine similarity
db_cos = FAISS.from_texts(texts, emb, distance_strategy=DistanceStrategy.COSINE)

# Euclidean distance
db_l2 = FAISS.from_texts(texts, emb, distance_strategy=DistanceStrategy.EUCLIDEAN)

# Max inner-product
db_ip = FAISS.from_texts(texts, emb, distance_strategy=DistanceStrategy.MAX_INNER_PRODUCT)


## Effective Prompt
#### Examples

In [56]:
from openai import OpenAI

# Initialize client
client = OpenAI()

# Mock RAG retrieved content
retrieved_regulation_chunks = """
Section 4.1: Transactions above $10,000 must be reported within 24 hours.
Section 5.3: All foreign transactions require dual authorization.
"""

# User question
audit_team_question = "Do we need to report domestic transactions above $15,000?"

# Construct Better Prompt
prompt = f"""
You are a Financial Compliance Assistant for internal audit teams.
Use ONLY the regulatory text provided in the CONTEXT section.
For the given QUESTION, follow these rules:

1. Provide a concise answer in 2–3 sentences.
2. Cite the specific regulation clause numbers or section titles you used.
3. If the regulation does not explicitly mention the requested information, state:
   "The regulation does not explicitly cover this point."
4. Do NOT guess or infer new rules. Do NOT provide advice beyond the context.

CONTEXT:
{retrieved_regulation_chunks}

QUESTION:
{audit_team_question}
"""

# Invoke LLM
response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt,
    max_output_tokens=250
)

# Output
print(response.output_text)


The regulation does not explicitly cover this point. Section 4.1 specifies reporting requirements for transactions above $10,000 but does not differentiate between domestic and foreign transactions.


In [57]:
from openai import OpenAI

client = OpenAI()

# Mock HR policy chunk (normally retrieved via RAG)
retrieved_hr_policy_chunks = """
Section 2.3: Employees can take up to 12 paid sick leaves per calendar year.
Section 4.1: Work-from-home is allowed for up to 8 days per month with manager approval.
Section 5.4: Overtime pay applies only when weekly work exceeds 40 hours.
"""

employee_question = "Am I allowed to work from home for 10 days in a month?"

prompt = f"""
You are an Internal HR Policy Assistant.
Use ONLY the HR policy text provided in the CONTEXT section.
For the given QUESTION, follow these rules:

1. Provide a clear answer in 2–4 sentences.
2. Cite the exact HR policy section numbers used.
3. If the policy does not explicitly cover the question, say:
   "The policy does not explicitly cover this issue."
4. Do NOT invent new HR rules or benefits.

CONTEXT:
{retrieved_hr_policy_chunks}

QUESTION:
{employee_question}
"""

response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt,
    max_output_tokens=200
)

print(response.output_text)


No, you are not allowed to work from home for 10 days in a month, as the policy permits work-from-home for up to 8 days per month with manager approval (Section 4.1).


In [58]:
from openai import OpenAI

client = OpenAI()

# Mock Security policy chunk (normally from vector DB)
retrieved_security_policy_chunks = """
Control ID SEC-12: All company laptops must use full-disk encryption.
Control ID SEC-18: Passwords must be changed every 90 days.
Control ID SEC-27: Multi-factor authentication (MFA) is mandatory for VPN access.
"""

it_security_question = "Is multi-factor authentication required for accessing company email?"

prompt = f"""
You are an Internal Information Security Policy Assistant.
Use ONLY the security policy text provided in the CONTEXT section.
For the given QUESTION, follow these rules:

1. Provide a precise answer in 2–4 sentences.
2. Cite the exact control IDs you used.
3. If the policy does not directly address the question, state:
   "The policy does not explicitly address this issue."
4. Do NOT invent new security requirements or controls.
5. Maintain compliance-ready, factual language.

CONTEXT:
{retrieved_security_policy_chunks}

QUESTION:
{it_security_question}
"""

response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt,
    max_output_tokens=200
)

print(response.output_text)


The policy does not explicitly address this issue.
